src/main/scala/firrtl/transforms/InlineBitExtractions.scala

```diff
@@ -0,0 +1,102 @@
+package firrtl
+package transforms
+
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.PrimOps.{Bits, Head, Tail, Shr}
+import firrtl.Utils.{isBitExtract, isTemp}
+import firrtl.WrappedExpression._
+
+import scala.collection.mutable
+
+object InlineBitExtractionsTransform {
+
+  // Checks if an Expression is made up of only Bits terminated by a Literal or Reference.
+  // private because it's not clear if this definition of "Simple Expression" would be useful elsewhere.
+  // Note that this can have false negatives but MUST NOT have false positives.
+  private def isSimpleExpr(expr: Expression): Boolean = expr match {
+    case _: WRef | _: Literal | _: WSubField => true
+    case DoPrim(op, args, _,_) if isBitExtract(op) => args.forall(isSimpleExpr)
+    case _ => false
+  }
+
+  // replace Head/Tail/Shr with Bits for easier back-to-back Bits Extractions
+  private def lowerToDoPrimOpBits(expr: Expression): Expression = expr match {
+    case DoPrim(Head, rhs, c, tpe) if isSimpleExpr(expr) && rhs.forall(isSimpleExpr) =>
+      val msb = bitWidth(rhs.head.tpe) - 1
+      val lsb = bitWidth(rhs.head.tpe) - c.head
+      DoPrim(Bits, rhs, Seq(msb,lsb), tpe)
+    case DoPrim(Tail, rhs, c, tpe) if isSimpleExpr(expr) && rhs.forall(isSimpleExpr) =>
+      val msb = bitWidth(rhs.head.tpe) - c.head - 1
+      DoPrim(Bits, rhs, Seq(msb,0), tpe)
+    case DoPrim(Shr, rhs, c, tpe) if isSimpleExpr(expr) && rhs.forall(isSimpleExpr) =>
+      DoPrim(Bits, rhs, Seq(bitWidth(rhs.head.tpe)-1, c.head), tpe)
+    case _ => expr // Not a candidate
+  }
+
+  /** Mapping from references to the [[firrtl.ir.Expression Expression]]s that drive them */
+  type Netlist = mutable.HashMap[WrappedExpression, Expression]
+
+  /** Recursively replace [[WRef]]s with new [[Expression]]s
+    *
+    * @param netlist a '''mutable''' HashMap mapping references to [[firrtl.ir.DefNode DefNode]]s to their connected
+    * [[firrtl.ir.Expression Expression]]s. It is '''not''' mutated in this function
```

Comment:
Is there a reason there is two `DefNode`s, e.g. `[[firrtl.ir.DefNode DefNode]]`. Same for `Expression`?

Reply:
no idea: here I just copy-pasted from InlineNots.scala, which was itself copy-pasted from ReplaceTruncatingArithmetic.scala and CombineCats.scala

Reply:
Ok, I presume this is a mistake, just remove the second duplicate word?

Reply:
This is so that you link to the first, but show the second. I can never ascertain when Scaladoc will properly resolve it (I think it's a function of the includes) and opt to be explicit with the full link, but not make the Scaladoc unreadable. Best bet is to run it through Scaladoc and see the output.

Reply:
Checking quickly... you need the `[[firrtl.ir.Expression Expression]]` here. Running this through `sbt unidoc`, you wind up with:

```
[warn] /home/schuyler/repos/github.com/freechipsproject/firrtl/src/main/scala/firrtl/transforms/InlineBitExtractions.scala:40:3: Could not find any member to link for "Expression".
[warn]   /** Recursively replace [[WRef]]s with new [[Expression]]s
```

Why Scaladoc can't find this is beyond me...